tests/sentry/api/bases/test_user.py

```diff
@@ -0,0 +1,80 @@
+import pytest
+from rest_framework.request import Request
+
+from sentry.api.bases.user import RegionSiloUserEndpoint, UserEndpoint, UserPermission
+from sentry.api.exceptions import ResourceDoesNotExist
+from sentry.testutils.cases import TestCase
+from sentry.testutils.silo import all_silo_test, control_silo_test, region_silo_test
+
+
+@all_silo_test(stable=True)
+class UserPermissionTest(TestCase):
+    user_permission = UserPermission()
+
+    def test_anonymous_no_user(self):
+        request = Request(self.make_request(method="GET"))
+        assert self.user_permission.has_object_permission(request, None, None)
+
+    def test_request_user(self):
+        user = self.create_user()
+        request = Request(self.make_request(method="GET"))
+        request.user = user
+        assert self.user_permission.has_object_permission(request, None, user)
+
+    def test_anonymous(self):
+        user = self.create_user()
+        request = Request(self.make_request(method="GET"))
+        assert not self.user_permission.has_object_permission(request, None, user)
+
+    def test_other(self):
+        request = Request(self.make_request(method="GET"))
+        request.user = self.create_user()
+        assert not self.user_permission.has_object_permission(request, None, self.create_user())
+
+
+@control_silo_test(stable=True)
+class UserEndpointTest(TestCase):
+    endpoint = UserEndpoint()
+
+    def test_retrieves_me_anonymous(self):
```

Comment:
I'd love to share these cases and parametrize on `{control,region}_silo_test` and `endpoint` but I couldn't figure out a way to make that work easily

Reply:
You could extract a base class that contains the tests, and then have two subclasses that apply either `control_silo_test` or `region_silo_test` 🤷

Reply:
I tried that initially and was having ...fun dealing with the type issues that created. On another glance I think it's actually easy to fix - thanks for the push in that direction!